### PR TITLE
Updated ISO generation for NUC 12

### PR DESCRIPTION
**Pull request by dsimog01**
Added necessary firmware for WiFi adapter in Intel NUC 12

### CODE REVIEW

**Comment by 3alpha** [comment]
We should also think about dropping support for older debian versions.

File: scripts/dappnode_install_pre.sh [hunk 1]
```diff
@@ -208,6 +246,13 @@ else
     install_lsof 2>&1 | tee -a $LOG_FILE
 fi
 
+# Only install wifi firmware if it does not detect the wifi card (Intel NUC 12)
+if [[ -z $(iw dev) ]]; then
```
Comment (3alpha) [inline scripts/dappnode_install_pre.sh:250]: There is no check that device is indeed NUC gen 12 and that it has WIFI card.

**Review by 3alpha — DISMISSED** [review]

**Comment by dsimog01** [comment]
I must check this to know if it is an Intel NUC 12 -> https://www.cyberciti.biz/faq/how-to-find-intel-nuc-bios-version-and-model-on-linux/

**Comment by dsimog01** [comment]
And I can check if installing certain packages from backports is enough, instead of upgrading

**Comment by 3alpha** [comment]
Take look at [this](https://wiki.debian.org/AptConfiguration) as well. It guides you how to pick specific packages from different repositories. We should use backports only where it is necessary.

File: scripts/dappnode_install_pre.sh (hunk 1 above)
Comment (dsimog01) [reply scripts/dappnode_install_pre.sh]: This has been solved in https://github.com/dappnode/DAppNode/pull/505/commits/bd4ba9e4605fe787c621acf52388f65e52b6ea04

File: scripts/dappnode_install_pre.sh [hunk 2]
```diff
@@ -208,6 +246,13 @@ else
...
+    echo -e "\e[32m \n\n Installing wifi firmware \n\n \e[0m" 2>&1 | tee -a $LOG_FILE
+    install_wifi_firmware 2>&1 | tee -a $LOG_FILE
+    upgrade_via_backports 2>&1 | tee -a $LOG_FILE
```
[the remark this answers is not in the document]
Comment (dsimog01) [reply scripts/dappnode_install_pre.sh]: I avoided upgrading from backports, but it is necessary to install linux-image-amd64 from backports for the wireless card to be detected (it also fixes screen rendering). Done here: https://github.com/dappnode/DAppNode/pull/505/commits/b42c3a811e6c1da16838a8a400eb755d28aef787

File: scripts/dappnode_install_pre.sh [hunk 3]
```diff
@@ -116,6 +117,43 @@ install_docker_compose() {
...
+    fi
+
+    # STEP 3: Clean up
+    rm -rf $TMP_FIRMWARE_DIR
+}
+
+# UPGRADE VIA BACKPORTS FOR WIFI FIRMWARE (For intel NUC 12)
+upgrade_via_backports() {
+    echo -e "deb http://deb.debian.org/debian bullseye-backports main contrib non-free" >> /etc/apt/sources.list
+    echo -e "deb-src http://deb.debian.org/debian bullseye-backports main contrib non-free" >> /etc/apt/sources.list
```
[the remark this answers is not in the document]
Comment (dsimog01) [reply scripts/dappnode_install_pre.sh]: Done here: https://github.com/dappnode/DAppNode/pull/505/commits/fff5d7e05f2c1872b97fec2a8065d71ae6210c05

**Comment by alexpeterson91** [comment]
> @dsimog01 great work!
> 
> Maybe we should consider creating a new ISO asset (to be implemented in the CI) just for the intel NUC 12, since this backports thingy is going to be temporary: `DAppNode-v0.2.59-debian-bullseye-amd64-unattended-intel-12.iso`
> 
> @3alpha @alexpeterson91 thoughts?

Yup i think that makes sense, at the same time we should just make a cloud ISO with IPFS set to server profile, its just chaning the IPFS .yaml file right?  and is the backports gonna be temporary? before this PR was done and i was just researching for a couple weeks i found several just linux questions about these firmwares and answers said due to bullseye's life cycle it may not be for a while that full support is availible.  similar issues in ubuntu had the same answers, dont expect anything for a while. like years, but idk just something i read a few times in a few places about multiple distros including debian bullseye

**Review by pablomendezroyo — APPROVED** [review]

**Review by pablomendezroyo — DISMISSED** [review]

**Review by pablomendezroyo — APPROVED** [review]

**Review by pablomendezroyo — CHANGES_REQUESTED** [review]
resolve conflicts

**Comment by alexpeterson91** [comment]
> We should also think about dropping support for older debian versions.

probably should be its own issue as it doesnt make sense in this PR, 

which needs to be added as like a separate branch and should be included in releases as a release asset until NUC12s are in our store or more ubiquitous (we've got to prep for the NUC13 too, its been out a month or 2 now and someones gonna buy it, i would have if i didnt get a 12 like 3 weeks before intel released Raptor Lake the 13th gen.)

**Comment by alexpeterson91** [comment]
> resolve conflicts

i know this isnt getting merged yet i just want to confirm conflicts are resolved currently as your review requested.

**Review by pablomendezroyo — DISMISSED** [review]

**Comment by tonoli** [comment]
Is there any update on this? 

How can I create an .iso using this branch before it's merged?

**Comment by alexpeterson91** [comment]
> Is there any update on this? 

I will update the branch to the latest core from this week and make sure the iso for NUC12 firmwares etc is published. I have to reinstall my own NUC12 anyway so I'll do final tests before it is published.

**Comment by alexpeterson91** [comment]
> Is there any update on this? 

If you want to build on your own. Clone the dappnode/dappnode repo (this one). If you have the SDK installed along with docker etc you can generate the ISO yourself by using CLI to enter the cloned repo the navigate into the ISO directory and then run `docker compose up` this will begin the iso generation process and will create a directory called images and will place the finished ISO there. I 
Still need to confirm it's all working on our end with the latest updates etc. which is on my list for today.

**Comment by tonoli** [comment]
> > Is there any update on this?
> 
> If you want to build on your own. Clone the dappnode/dappnode repo (this one). If you have the SDK installed along with docker etc you can generate the ISO yourself by using CLI to enter the cloned repo the navigate into the ISO directory and then run `docker compose up` this will begin the iso generation process and will create a directory called images and will place the finished ISO there. I Still need to confirm it's all working on our end with the latest updates etc. which is on my list for today.

Thank you very much for you help! 

I managed to install by doing that but unfortunately the dappnode then has no internet connection through ethernet or wifi.

I did this: 
- Cloned the repo 
- Then ran the commands: 
`docker-compose build`
`docker-compose up`
- flashed the generated iso
- managed to solve the installation problem of Debian not detecting the network material. 
- All the installation went smoothly the dappnode runs well but the ethernet connection doesn't work

**Comment by alexpeterson91** [comment]
@pablomendezroyo @dsimog01 ive been checking in on this PR/branch since D started having me test it, and have been updating the branch for each core release and any other change It currently passes all CI checks, has no conflicts, but still hasnt been merged.  I Know Pablo you suggested releasing it as a standalone asset with each release, since its temporary and will be replaced/ improved in the near future.  However we have more and more users with NUC12s at home and no way to get them working with our current release assets they're coming up with some messy workarounds, we've had this working for months, this branch's ISOs should be published for the next core release as an additional asset if the backports check is too much, and we should upload a NUC12 compatible ISO for the current release as soon as possible.

**Comment by dsimog01** [comment]
@alexpeterson91 There is a newer Debian version (11.6). I will test if it avoids updating linux-image from backports

**Comment by alexpeterson91** [comment]
I can test Too my NUc 12 has been out of commission since we were last testing. Mainly due to lack of knowing if it was still needed for more testing and also I needed more IPs I have too many over powered NUCs on ny home connection anyway it made more sense to shelve it till they finalized my  new internet service or I needed to test it again. Now I got my dedicated fiber activated this weekend and have several more high end test and production servers coming online very soon. My nuc 12 will likely be the first device active on my new setup. 
Just let me know if you need me to check anything on my machine it's sitting just waiting to be hooked back up and reinstalled.

**Comment by dsimog01** [comment]
> I can test Too my NUc 12 has been out of commission since we were last testing. Mainly due to lack of knowing if it was still needed for more testing and also I needed more IPs I have too many over powered NUCs on ny home connection anyway it made more sense to shelve it till they finalized my new internet service or I needed to test it again. Now I got my dedicated fiber activated this weekend and have several more high end test and production servers coming online very soon. My nuc 12 will likely be the first device active on my new setup. Just let me know if you need me to check anything on my machine it's sitting just waiting to be hooked back up and reinstalled.

I already tested updating to the latest Debian release (11.6) and it does not fix the issue we are experiencing with NUC 12s, so we still need to update linux image from backports to make it work

**Comment by alexpeterson91** [comment]
> > I can test Too my NUc 12 has been out of commission since we were last testing. Mainly due to lack of knowing if it was still needed for more testing and also I needed more IPs I have too many over powered NUCs on ny home connection anyway it made more sense to shelve it till they finalized my new internet service or I needed to test it again. Now I got my dedicated fiber activated this weekend and have several more high end test and production servers coming online very soon. My nuc 12 will likely be the first device active on my new setup. Just let me know if you need me to check anything on my machine it's sitting just waiting to be hooked back up and reinstalled.
> 
> I already tested updating to the latest Debian release (11.6) and it does not fix the issue we are experiencing with NUC 12s, so we still need to update linux image from backports to make it work

Well then we need to polish this and add it as a standalone asset for download for releases until its no longer an issue, which feels to me like it could be longer than we want.  Users are struggling, some 11s, all 12s, and 13s all have this issue missing the drivers and firmware for the essential components like LAN and WiFi.  Right? @pablomendezroyo  your thoughts? you suggested it a while ago earlier in this PR, i think its what needs to be done agreed then, and since we still have to deal with backports for now we should just keep the branches on updated to each other save for the changes in this branch which we just publish as a standalone release asset for the current core, and the upcoming versions until its no longer needed.

 its a lot of users running into this ow, and some ways people are doing it is really messy and ive tried to tell people not to share those messy methods as they may mess up our normal upgrades and services; as most involve updating and upgrading system files over sh or direct console access with keyboard. I really just wanna get the core versions updated and publish a new ISO for NUC12+ devices since we will only encounter more and more every day

**Comment by alexpeterson91** [comment]
> > > > I can test Too my NUc 12 has been out of commission since we were last testing. Mainly due to lack of knowing if it was still needed for more testing and also I needed more IPs I have too many over powered NUCs on ny home connection anyway it made more sense to shelve it till they finalized my new internet service or I needed to test it again. Now I got my dedicated fiber activated this weekend and have several more high end test and production servers coming online very soon. My nuc 12 will likely be the first device active on my new setup. Just let me know if you need me to check anything on my machine it's sitting just waiting to be hooked back up and reinstalled.
> 
> > > 
> 
> > > 
> 
> > > I already tested updating to the latest Debian release (11.6) and it does not fix the issue we are experiencing with NUC 12s, so we still need to update linux image from backports to make it work
> 
> > 
> 
> > Well then we need to polish this and add it as a standalone asset for download for releases until its no longer an issue, which feels to me like it could be longer than we want. Users are struggling, some 11s, all 12s, and 13s all have this issue missing the drivers and firmware for the essential components like LAN and WiFi. Right? @pablomendezroyo your thoughts? you suggested it a while ago earlier in this PR, i think its what needs to be done agreed then, and since we still have to deal with backports for now we should just keep the branches on updated to each other save for the changes in this branch which we just publish as a standalone release asset for the current core, and the upcoming versions until its no longer needed.
> 
> > 
> 
> > its a lot of users running into this ow, and some ways people are doing it is really messy and ive tried to tell people not to share those messy methods as they may mess up our normal upgrades and services; as most involve updating and upgrading system files over sh or direct console access with keyboard. I really just wanna get the core versions updated and publish a new ISO for NUC12+ devices since we will only encounter more and more every day
> 
> 
> 
> Hello, I run into the NUC12+ problem as well, I tried to compose the iso from this branch, but for some reason, the installation process always failed, how do remove all the messy failed files and restart the installation process, is it simply restart the device with the USB? But it never let me set the user id and password, feel like something is missing.

We will be releasing a prebuilt experimental ISO compatible with NUC11 extremes, and NUC12s and 13s shortly. It's been tested a lot but not long term. The issue holding us back right now is that we need a newer kernel to make the drivers for the various peripherals like the NIC cards and WiFi audio drivers on some models and PCIe drivers on top end models. Debian doesn't have this available just yet in a stable form. Debian 12 is what we need to make this workaround install (which works, but is not reliable over the longer term). So it will be an experimental asset and to be used your own risk. 
It's a stopgap until Debian 12 is released in March as of the last time I checked. They just completed their first freeze on the release so it's not too far away but this "experimental" ISO release will be made available very shortly as a stop gap until Debian 12 is released as stable.

**Comment by dsimog01** [comment]
It seems like we do not need this anymore as it looks like Debian 12 (https://github.com/dappnode/DAppNode/pull/575) is solving the issues we were having that made this PR necessary.

**Pull request closed without merging**